pkg/orchestrator/discovery_worker.go

```diff
@@ -19,16 +20,43 @@ func (n *DiscoveryWorker) Run(work interface{}) error {
 
 			log.Printf("Found %d applications for integration provider %s.", len(applications), p.Name())
 			for _, app := range applications {
-				_, err := n.Gateway.CreateIfAbsent(record.ID, app.ObjectID, app.Name, app.Description, app.Service) // idempotent work
+				id, err := n.Gateway.CreateIfAbsent(record.ID, app.ObjectID, app.Name, app.Description, app.Service) // idempotent work
 				if err != nil {
 					log.Printf("Failed to create application: %s", err)
+					continue
 				}
+				discoveredApps[id] = struct{}{}
 			}
 		}
 	}
+
+	if len(discoveredApps) > 0 {
+		n.removeUnknownApps(discoveredApps)
+	}
 	return nil
 }
 
+func (n *DiscoveryWorker) removeUnknownApps(discoveredApps map[string]struct{}) {
+	allApps, err := n.Gateway.Find()
+	if err != nil {
+		log.Printf("Failed to get applications: %s", err)
+		return
+	}
+	appsToDelete := make([]string, 0)
+	for _, app := range allApps {
+		if _, found := discoveredApps[app.ID]; !found {
+			appsToDelete = append(appsToDelete, app.ID)
+		}
+	}
+
+	for _, id := range appsToDelete {
+		derr := n.Gateway.DeleteById(id)
```

Comment:
I was thinking of changing this signature to be variadic. `n.Gateway.DeleteById(appsToDelete...)`.

The thought here was to make a single database query instead of having to loop and make multiple queries. Since it's not that much of "performance" hit I decided to default to keeping it simple and just deleting one id at a time.